### ISSUE TITLE
Examples?

### ISSUE THREAD

**Issue by isaachess** [pixelbender/go-stun#6]
I'm struggling to use this effectively and am wondering if you could provide some examples.

The actually communication with the STUN server works fine: it returns the public addr observed by the STUN server.

But USING this information isn't working for me, and it has me wondering if I'm just misunderstanding how the package works (thus the request for examples).

For example, when calling `stun.Discover` it returns a conn and addr. I'm attempting, after performing the discover, to receive UDP packets sent to addr. So I'm trying to read from the return conn (while sending UDP messages to `addr` from another machine), but nothing ever comes through.

The purpose of STUN in my understanding is to hole-punch so peers can send UDP traffic to the public `addr`, which is then forwarded to the correct host. Can I see an example of how this package gets that done?

**Comment by pixelbender** [comment]
@isaachess Yes, It should work exactly you described, but currently stun.Discover uses net.Dial instead of Listen (so, the socket is associated with a STUN-server's address) and hole-punching does not work. I'll fix it, thanks.

**Comment by kenkeiter** [comment]
I just spent the better part of a day tearing my hair out over this, so I wanted to make sure that I documented exactly what's happening here, in the hopes that someone else won't have to.

Although @pixelbender fixed the issue with the fact that `net.Listen*` wasn't being used, the `net.PacketConn` returned by `Discover()` isn't _actually_  relinquished in a way that allows it to be repurposed! When the `Conn` is created, an `Agent` is [set up](https://github.com/pixelbender/go-stun/blob/master/stun/conn.go#L16) to handle it. [Upon start](https://github.com/pixelbender/go-stun/blob/master/stun/agent.go#L92-L94), the `Agent` calls [this method](https://github.com/pixelbender/go-stun/blob/master/stun/agent.go#L127-L141), which calls `ReadFrom` on the `net.PacketConn` to loop over/handle received messages. Turns out, when the `net.PacketConn` is [handed back to you](https://github.com/pixelbender/go-stun/blob/master/stun/stun.go#L22) from `Discover()`, that loop is never actually stopped.

I guess that's what @pixelbender meant by the `TODO: hijack` comment 😜 Anyway, long story short, it turns out that this issue hasn't really been addressed.

**Comment by kenkeiter** [comment]
@isaachess That PR (#8) should allow you to use this library for hole-punching. The `net.PacketConn` you get back is now usable, so if you know the public address (host/port) of your peer, you can get them to try to send packets to one another at the same time.

For what it's worth, you'll do something _like_ this for both clients (although you still need an external service to help them exchange addresses):

```go
conn, externalAddr, err := stun.Discover(STUNServer)
if err != nil {
	return nil, errors.Wrap(err, "failed to set up public UDP socket")
}
peerExternalAddr := &net.UDPAddr{
	IP: net.ParseIP("<peer addr>"),
	Port: 1234,
}
conn.WriteTo([]byte("hello"), peerExternalAddr)
```

**Comment by mennanov** [comment]
Could you please point me to the example on how to setup a TCP connection between 2 nodes both behind NAT?
Something like node1_code.go, node2_code.go, external_server_code.go
Thanks!
